module-system/sov-state/src/vec.rs

```diff
@@ -0,0 +1,226 @@
+    /// To push a value to the end of the StateVec, use [`StateVec::push`].
+    pub fn set<S: Storage>(
+        &self,
+        index: usize,
+        value: &V,
+        working_set: &mut WorkingSet<S>,
+    ) -> Result<(), Error> {
+        let len = self.len(working_set);
+
+        if index < len {
+            working_set.set_value(
+                self.prefix(),
+                &self.internal_codec(),
+                &IndexKey(index + 1),
+                value,
+            );
+            Ok(())
+        } else {
+            Err(Error::IndexOutOfBounds(index))
+        }
+    }
+
+    /// Returns the value for the given index.
+    pub fn get<S: Storage>(&self, index: usize, working_set: &mut WorkingSet<S>) -> Option<V> {
+        working_set.get_value(self.prefix(), &self.internal_codec(), &IndexKey(index + 1))
+    }
+
+    /// Returns the value for the given index.
+    /// If the index is out of bounds, returns an error.
+    /// If the value is absent, returns an error.
+    pub fn get_or_err<S: Storage>(
+        &self,
+        index: usize,
+        working_set: &mut WorkingSet<S>,
+    ) -> Result<Option<V>, Error> {
+        let len = self.len(working_set);
+
+        if index < len {
+            let elem =
+                working_set.get_value(self.prefix(), &self.internal_codec(), &IndexKey(index + 1));
+
+            if elem.is_some() {
+                Ok(elem)
+            } else {
+                Err(Error::MissingValue(self.prefix().clone(), index))
+            }
+        } else {
+            Err(Error::IndexOutOfBounds(index))
+        }
+    }
+
+    /// Returns the length of the [`StateVec`].
+    pub fn len<S: Storage>(&self, working_set: &mut WorkingSet<S>) -> usize {
+        let len = working_set.get_value::<_, usize, _>(
+            self.prefix(),
+            &self.internal_codec(),
+            &IndexKey(0),
+        );
+        len.unwrap_or_default()
+    }
+
+    /// Pushes a value to the end of the [`StateVec`].
+    pub fn push<S: Storage>(&self, value: &V, working_set: &mut WorkingSet<S>) {
+        let len = self.len(working_set);
+
+        working_set.set_value(
+            self.prefix(),
+            &self.internal_codec(),
+            &IndexKey(len + 1),
+            value,
+        );
+        self.set_len(len + 1, working_set);
+    }
+
+    /// Pops a value from the end of the [`StateVec`] and returns it.
+    pub fn pop<S: Storage>(&self, working_set: &mut WorkingSet<S>) -> Option<V> {
+        let len = self.len(working_set);
+
+        if len > 0 {
+            let elem =
+                working_set.remove_value(self.prefix(), &self.internal_codec(), &IndexKey(len));
+            self.set_len(len - 1, working_set);
+            elem
+        } else {
+            None
+        }
+    }
+
+    /// Sets all values in the [`StateVec`].
+    /// If the length of the provided values is less than the length of the [`StateVec`], the remaining values stay in storage but are inaccessible.
```

Comment:
> /// If the length of the provided values is less than the length of the [`StateVec`], the remaining values stay in storage but are inaccessible.

This is a bit of an odd feature - in general, you really don't want the state trie getting filled up with inaccessible values. What's the motivating use case for this behavior?

Reply:
Doesn't performance benefit from this pattern? This way you're not paying the price of removing all that data.

Reply:
Performance does benefit for the person clearing the map, but at the expense of everyone else's performance forever. Those stale values still have to be stored by all full nodes, and they fill up the JMT making merkle paths longer 😬